Enable tests to export unique keys

idorax opened the pull request:
`tmt tests export` prints duplicate keys if multiple tests are exported. To fix it, we should create a list for tests
data before exporting.

happz commented on tmt/base.py:
```diff
@@ -536,22 +535,17 @@ def export(self, format_='yaml', keys=None):
         # Common node export otherwise
-        elif keys == 'fmf-id':
-            if format_ == 'dict':
+        elif formate_ == 'dict' or formate_ == 'yaml':
```
This seems to be a typo, `formate_` vs `format_`.

idorax replied:
thank you, @happz, the typo is fixed.

idorax commented on tmt/base.py:
```diff
@@ -535,23 +535,22 @@ def export(self, format_='yaml', keys=None):
-        elif format_ == 'nitrate':
-            tmt.export.export_to_nitrate(self)
-
+            return (name, data)
```
OK, it's reverted.

psss commented on tmt/base.py:
```diff
@@ -542,16 +542,18 @@ def export(self, format_='yaml', keys=None):
         # Common node export otherwise
-        elif keys == 'fmf-id':
-            if format_ == 'dict':
+        elif format_ == 'dict' or format_ == "yaml":
```
Current change causes that `Test.export()` returns a `dict` when asked for `yaml`. For example:

    >>> test = tmt.Tree('.').tests()[0]
    >>> test.export(format_='yaml')
    {'name': '/tests/clean/guests', 'summary': 'Checks whether tmt clean...

This is not expected and could be very confusing behaviour.

idorax replied:
How about `Plan.export()` return if we asked for `yaml`? e.g.
```
>>> plan = tmt.Tree('.').plans()[0]
>>> plan.export(format_='yaml')
```

idorax replied:
With patch in the following,
```patch
diff --git a/tmt/base.py b/tmt/base.py
index 5457b10..cba5594 100644
--- a/tmt/base.py
+++ b/tmt/base.py
@@ -550,7 +550,10 @@ class Test(Core):
                 data['name'] = self.name
                 data.update(super().export(format_='dict'))
                 data.pop('adjust', None)
-                return data
+                if format_ == 'dict':
+                    return data
+                else:
+                    return tmt.utils.dict_to_yaml(data)
         else:
             raise tmt.utils.GeneralError(
                 f"Invalid test export format '{format_}'.")
diff --git a/tmt/cli.py b/tmt/cli.py
index e8875f5..3393624 100644
--- a/tmt/cli.py
+++ b/tmt/cli.py
@@ -581,15 +581,12 @@ def export(context, format_, nitrate, bugzilla, **kwargs):
         if kwargs.get('fmf_id'):
             keys = 'fmf-id'
 
-        data = list()
-        for test in context.obj.tree.tests():
-            test_info = test.export(format_=format_, keys=keys)
-            data.append(test_info)
-
+        tests = [test.export(format_='dict', keys=keys) for test in
+            context.obj.tree.tests()]
         if format_ == 'dict':
-            echo(data, nl=False)
+            echo(tests, nl=False)
         else:
-            echo(tmt.utils.dict_to_yaml(data), nl=False)
+            echo(tmt.utils.dict_to_yaml(tests), nl=False)
     else:
         raise tmt.utils.GeneralError(
             f"Invalid test export format '{format_}'.")
```

`test.export(format_='yaml')` works fine now. 
```python
>>> test = tmt.Tree('.').tests()[0]
>>> test.export(format_='yaml')
'name: /tests/example1\nsummary: Test01\ndescription:\ncontact: []\ncomponent: []\ntest: ./test.sh\npath: /tests/example1\nframework: beakerlib\nmanual: false\nrequire:\n  - beakerlib\nrecommend: []\nenvironment: {}\nduration: 5m\nenabled: true\norder: 50\nresult: respect\ntag: []\ntier:\nlink: []\n'
>>>
>>> test.export(format_='dict')
{'name': '/tests/example1', 'summary': 'Test01', 'description': None, 'contact': [], 'component': [], 'test': './test.sh', 'path': '/tests/example1', 'framework': 'beakerlib', 'manual': False, 'require': ['beakerlib'], 'recommend': [], 'environment': {}, 'duration': '5m', 'enabled': True, 'order': 50, 'result': 'respect', 'tag': [], 'tier': None, 'link': []}
>>> 
>>> 
>>> plan = tmt.Tree('.').plans()[0]
>>> plan.export(format_='yaml')
'name: /plans/example1\nsummary: Basic smoke test 01\ndescription:\nenabled: true\norder: 50\nlink: []\ndiscover:\n    how: fmf\n    name: default\nexecute:\n    how: tmt\n    name: default\n'
>>>
>>> plan.export(format_='dict')
{'name': '/plans/example1', 'summary': 'Basic smoke test 01', 'description': None, 'enabled': True, 'order': 50, 'link': [], 'discover': {'how': 'fmf', 'name': 'default'}, 'execute': {'how': 'tmt', 'name': 'default'}}
```

psss replied:
Yes, now works as expected. No need to include patches in the comments. Just add the fix as a separate commit. If needed, you can then easily reference it using the commit hash.

psss commented on tmt/base.py:
```diff
@@ -542,16 +542,18 @@ def export(self, format_='yaml', keys=None):
-                raise tmt.utils.GeneralError(
-                    f"Invalid test export format '{format_}'.")
+                data = {}
+                data['name'] = self.name
```
Instead of adding the `name` key here, let's update the `Core.export()` method instead to always include the `name` key unless an explicit list of the desired `keys` is provided as the parameter. We can then remove similar action from the `Plan.export()` method as well and `Test.export()` will need almost no change.

psss replied:
Thanks.

psss commented on tmt/cli.py:
```diff
@@ -569,14 +569,30 @@ def export(context, format_, nitrate, bugzilla, **kwargs):
+    if nitrate:
+        for test in context.obj.tree.tests():
             test.export(format_='nitrate')
+    elif format_ == 'execute':
```
The `execute` format is not needed from the command line. It is just used by the `detach` executor (which btw will be removed soon).

psss replied:
Good thanks.

psss commented on tmt/cli.py:
```diff
@@ -569,14 +569,30 @@ def export(context, format_, nitrate, bugzilla, **kwargs):
+            echo(tmt.utils.dict_to_yaml(data), nl=False)
+    else:
+        raise tmt.utils.GeneralError(
+            f"Invalid test export format '{format_}'.")
```
I'd suggest to use the same approach as for `tmt plan export` below:

https://github.com/teemtee/tmt/blob/ec2e296a88bceba9eed5f14667cc45ad5d6dc5fc/tmt/cli.py#L720

Fetch all test data using `dict` format and then send to the user `dict` or `yaml`. I believe the implementation should then be much more straightforward.

idorax replied:
Yes, updated.

psss reviewed (changes requested):
Thanks for looking into this. I've proposed several improvemnts which should make the code a bit more straightforward. Could you please also include at least a simple test coverage? I'd suggest to start with creating a `beakerlib` template:

    tmt test create -t beakerlib /tests/test/export

In `beakerlib` there are some useful features like running command and checking the output or making nice headers which you are implementing in your `unit-test-run.sh` script. You can also get some inspiration from other tests, e.g. `/tests/test/select`.

psss commented:
> The `name` issue is fixed with patch in the following,

Good thanks. Just a hint: When working on the pull request (especially when addressing review comments) it's better to add new commits on the top of your changes instead of amending and force-pushing. It makes the review easier. See also [pull request](https://tmt.readthedocs.io/en/latest/contribute.html#pull-requests) recommendations.

> I'm working on it.

Great!

> BTW, how can I run a single test via tmt?

Just add `test --name NAME` or `test -n NAME` to the `tmt run` command:

    tmt run -v test -n plan/export

It might make sense to choose the plan as well (to skip other plans):

    tmt run -v test -n plan/export plan -n basic

The fastest way (if you don't need any provision or preparation) is to run the test directly:

    cd tests/plan/export/cd tests/plan/export/
    ./test.sh

Hope this helps.

psss commented:
> $ tmt test ls | egrep 'export'

A shorter version is `tmt test ls export` ;-)

idorax commented:
> > The `name` issue is fixed with patch in the following,
> 
> Good thanks. Just a hint: When working on the pull request (especially when addressing review comments) it's better to add new commits on the top of your changes instead of amending and force-pushing. It makes the review easier. See also [pull request](https://tmt.readthedocs.io/en/latest/contribute.html#pull-requests) recommendations.

I didn't know multiple commits in a PR will be merged into one commit when the PR is merged, so I always use `--amend` before doing `git push -f` during the code review phase. @psss, will you help to merge multiple commits into a single commit when you merge a PR?

idorax commented:
Hi @psss, the test coverage for `tmt test export` is added, please help to review, thanks!

psss commented on tests/test/export/data/tests.fmf:
```diff
@@ -0,0 +1,31 @@
+        tier: 1
+    /two:
+        summary: Explicit tier level
+        tier: 2
```
Why are you including 9 tests and then checking them all in the very same way? Wouldn't two or three tests be just enough? Doing basically the same checks does not bring any benefit, just makes the test slower.

idorax replied:
I just copied it from tests/test/select/data/tests.fmf. Now there are only two cases.

```
test: 'true'

/enabled:
    /default:
        summary: This test is enabled by default
    /disabled:
        summary: This test is disabled
        enabled: false
```

psss replied:
Good, thanks!

psss commented on tests/test/export/main.fmf:
```diff
@@ -0,0 +1,4 @@
+summary: Test the feature to export tests to yaml or dict
+contact: Vector Li <huanli@redhat.com>
+test: ./test.sh
+framework: beakerlib
```
The last two lines are not necessary, they are inherited from here:

https://github.com/teemtee/tmt/blob/4272d242da23f3d5dd5be8db1f683720c22ac72f/tests/main.fmf#L1

idorax replied:
Removed.

psss replied:
Thanks.

psss commented on tests/test/export/test.sh:
```diff
@@ -0,0 +1,145 @@
+    tnames+=" /tests/test/export/data/tests/tag/empty"
+    tnames+=" /tests/test/export/data/tests/tier/default"
+    tnames+=" /tests/test/export/data/tests/tier/one"
+    tnames+=" /tests/test/export/data/tests/tier/two"
```
Please, do a `tmt init` in the `/tests/test/export/data` directory to mark root of a new `fmf.Tree`. The `data` directory should include its own `.fmf` directory so that the export test data do not overlap/mix with the main fmf metadata.

idorax replied:
Done.

psss commented on tests/test/export/test.sh:
```diff
@@ -0,0 +1,145 @@
+            rlAssertGrep "- name: $tname" $rlRun_LOG
+        done
+        verify_output "yaml" "$fields"
+    rlPhaseEnd
```
This grepping of individual lines does not actually verify that a valid list of multiple tests is actually exported. Except for the missing `name` the lines above pass without any problem with the old broken implementation as well.

What about doing a real parsing to verify that the output it really a valid `yaml`? It could look like this:

```shell
rlRun -s "tmt tests export | ./parse.py"
```

```python
#!/usr/bin/python3

import sys
import ruamel.yaml

yaml = ruamel.yaml.YAML()
tests = yaml.load(sys.stdin)

print(tests[0]["name"])
assert tests[0]["name"] == "/test/one"

print(tests[1]["name"])
assert tests[0]["name"] == "/test/two"

print(tests[2]["name"])
assert tests[0]["name"] == "/test/three"
```

This would actually catch the problem with the original output:

```
found duplicate key "summary" with value "This test is enabled explicitly" (original value: "This test is enabled by default")
  in "<stdin>", line 19, column 1
```

What do you think?

idorax replied:
@psss, nice suggestion! I have created the [parse.py][01] file, please help to review again, thanks!

[01]: https://github.com/idorax/tmt/blob/dev.huanli.20220321.1091.export/tests/test/export/parse.py

psss reviewed (changes requested):
Thanks for writing the test. Added a couple of comments & questions.

idorax commented on tests/test/export/test.sh:
```diff
@@ -0,0 +1,145 @@
+        rlPhaseEnd
+    done
+
+    rlPhaseStartTest "fmf-id not exsist"
```
Updated.

psss reviewed (approved):
Thanks for the improvements. Looks good. I've just fixed `yaml` export of `fmf` identifiers and added a couple of minor adjustments in 2c029ff. Rebased and squashed. Let's see if tests are green...

idorax commented:
> Thanks for the improvements. Looks good. I've just fixed `yaml` export of `fmf` identifiers and added a couple of minor adjustments in [2c029ff](https://github.com/teemtee/tmt/commit/2c029ffe3c37b4a88845ac5092e7f50b913fae9e). Rebased and squashed. Let's see if tests are green...

Looks good to me, thanks! @psss
